File: app/kotlin/puyo-puyo-tdd/src/commonTest/kotlin/StageTest.kt
package com.example.puyopuyo

import kotlin.test.Test
import kotlin.test.assertEquals

class StageTest {
    @Test
    fun 初期状態では全てのマスが空である() {
        // Arrange
        val config = Config()
        val stage = Stage(config)
        stage.initialize()

        // Act & Assert
        for (y in 0 until config.stageHeight) {
            for (x in 0 until config.stageWidth) {
                assertEquals(0, stage.getPuyo(x, y))
            }
        }
    }

    @Test
    fun ぷよを配置できる() {
        // Arrange
        val config = Config()
        val stage = Stage(config)
        stage.initialize()

        // Act
        stage.setPuyo(2, 10, 1) // (x=2, y=10) に種類1のぷよを配置

        // Assert
        assertEquals(1, stage.getPuyo(2, 10))
    }

    @Test
    fun 複数のぷよを配置できる() {
        // Arrange
        val config = Config()
        val stage = Stage(config)
        stage.initialize()

        // Act
        stage.setPuyo(0, 12, 1) // 赤ぷよ
        stage.setPuyo(1, 12, 2) // 青ぷよ
        stage.setPuyo(2, 12, 3) // 緑ぷよ
        stage.setPuyo(3, 12, 4) // 黄ぷよ

        // Assert
        assertEquals(1, stage.getPuyo(0, 12))
        assertEquals(2, stage.getPuyo(1, 12))
        assertEquals(3, stage.getPuyo(2, 12))
        assertEquals(4, stage.getPuyo(3, 12))
    }

    @Test
    fun ぷよを上書きできる() {
        // Arrange
        val config = Config()
        val stage = Stage(config)
        stage.initialize()
        stage.setPuyo(2, 10, 1)

        // Act
        stage.setPuyo(2, 10, 2) // 上書き

        // Assert
        assertEquals(2, stage.getPuyo(2, 10))
    }

    @Test
    fun 隣接する同じ色のぷよを探索できる() {
        // Arrange
        val config = Config()
        val stage = Stage(config)
        stage.initialize()

        // 縦に4つ並べる
        stage.setPuyo(2, 9, 1)
        stage.setPuyo(2, 10, 1)
        stage.setPuyo(2, 11, 1)
        stage.setPuyo(2, 12, 1)

        // Act
        val connectedPuyos = stage.searchConnectedPuyo(2, 9, 1)

        // Assert
        assertEquals(4, connectedPuyos.size)
    }

    @Test
    fun L字型につながったぷよを探索できる() {
        // Arrange
        val config = Config()
        val stage = Stage(config)
        stage.initialize()

        // L字型に配置
        stage.setPuyo(2, 11, 1)
        stage.setPuyo(2, 12, 1)
        stage.setPuyo(3, 12, 1)
        stage.setPuyo(4, 12, 1)

        // Act
        val connectedPuyos = stage.searchConnectedPuyo(2, 11, 1)

        // Assert
        assertEquals(4, connectedPuyos.size)
    }
}

File: app/csharp-2/wpf/tests/PuyoPuyo.Tests/Models/StageTest.cs
using Xunit;
using PuyoPuyo.WPF.Models;

namespace PuyoPuyo.Tests.Models;

public class StageTest
{
    private readonly Config _config;
    private readonly Stage _stage;

    public StageTest()
    {
        _config = new Config();
        _stage = new Stage(_config.StageWidth, _config.StageHeight);
    }

    [Fact]
    public void 同じ色のぷよが4つつながっていると消去対象になる()
    {
        // Arrange
        _stage.Clear();
        _stage.SetCell(1, 10, 1);
        _stage.SetCell(2, 10, 1);
        _stage.SetCell(1, 11, 1);
        _stage.SetCell(2, 11, 1);

        // Act
        var eraseInfo = _stage.CheckErase();

        // Assert
        Assert.Equal(4, eraseInfo.ErasePuyoCount);
        Assert.NotEmpty(eraseInfo.EraseList);
    }

    [Fact]
    public void 同じ色のぷよが3つでは消去対象にならない()
    {
        // Arrange
        _stage.Clear();
        _stage.SetCell(1, 10, 1);
        _stage.SetCell(2, 10, 1);
        _stage.SetCell(1, 11, 1);

        // Act
        var eraseInfo = _stage.CheckErase();

        // Assert
        Assert.Equal(0, eraseInfo.ErasePuyoCount);
        Assert.Empty(eraseInfo.EraseList);
    }

    [Fact]
    public void 異なる色のぷよは消去対象にならない()
    {
        // Arrange
        _stage.Clear();
        _stage.SetCell(1, 10, 1);
        _stage.SetCell(2, 10, 2);
        _stage.SetCell(1, 11, 2);
        _stage.SetCell(2, 11, 1);

        // Act
        var eraseInfo = _stage.CheckErase();

        // Assert
        Assert.Equal(0, eraseInfo.ErasePuyoCount);
        Assert.Empty(eraseInfo.EraseList);
    }

    [Fact]
    public void 縦に5つつながっていても消去できる()
    {
        // Arrange
        _stage.Clear();
        _stage.SetCell(2, 7, 1);
        _stage.SetCell(2, 8, 1);
        _stage.SetCell(2, 9, 1);
        _stage.SetCell(2, 10, 1);
        _stage.SetCell(2, 11, 1);

        // Act
        var eraseInfo = _stage.CheckErase();

        // Assert
        Assert.Equal(5, eraseInfo.ErasePuyoCount);
    }

    [Fact]
    public void L字型につながっていても消去できる()
    {
        // Arrange
        _stage.Clear();
        _stage.SetCell(1, 10, 1);
        _stage.SetCell(1, 11, 1);
        _stage.SetCell(2, 11, 1);
        _stage.SetCell(3, 11, 1);

        // Act
        var eraseInfo = _stage.CheckErase();

        // Assert
        Assert.Equal(4, eraseInfo.ErasePuyoCount);
    }

    [Fact]
    public void 消去情報に基づいてぷよを消去できる()
    {
        // Arrange
        _stage.Clear();
        _stage.SetCell(1, 10, 1);
        _stage.SetCell(2, 10, 1);
        _stage.SetCell(1, 11, 1);
        _stage.SetCell(2, 11, 1);
        var eraseInfo = _stage.CheckErase();

        // Act
        _stage.Erase(eraseInfo);

        // Assert
        Assert.Equal(0, _stage.GetCell(1, 10));
        Assert.Equal(0, _stage.GetCell(2, 10));
        Assert.Equal(0, _stage.GetCell(1, 11));
        Assert.Equal(0, _stage.GetCell(2, 11));
    }

    [Fact]
    public void 空中に浮いているぷよは重力で落下する()
    {
        // Arrange
        _stage.Clear();
        _stage.SetCell(2, 5, 1); // 空中のぷよ
        _stage.SetCell(2, 11, 2); // 一番下のぷよ

        // Act
        _stage.ApplyGravity();

        // Assert
        Assert.Equal(0, _stage.GetCell(2, 5)); // 元の位置は空
        Assert.Equal(1, _stage.GetCell(2, 10)); // 下に落ちた
        Assert.Equal(2, _stage.GetCell(2, 11)); // 底は変わらず
    }

    [Fact]
    public void 複数の列で同時に落下処理が行われる()
    {
        // Arrange
        _stage.Clear();
        _stage.SetCell(1, 5, 1);
        _stage.SetCell(2, 6, 2);
        _stage.SetCell(3, 7, 3);

        // Act
        _stage.ApplyGravity();

        // Assert
        Assert.Equal(1, _stage.GetCell(1, 11));
        Assert.Equal(2, _stage.GetCell(2, 11));
        Assert.Equal(3, _stage.GetCell(3, 11));
    }

    [Fact]
    public void 下にぷよがあると落下が止まる()
    {
        // Arrange
        _stage.Clear();
        _stage.SetCell(2, 8, 1); // 上のぷよ
        _stage.SetCell(2, 10, 2); // 障害物
        _stage.SetCell(2, 11, 3); // 底

        // Act
        _stage.ApplyGravity();

        // Assert
        Assert.Equal(1, _stage.GetCell(2, 9)); // 障害物の上に落ちる
        Assert.Equal(2, _stage.GetCell(2, 10));
        Assert.Equal(3, _stage.GetCell(2, 11));
    }
}
